File: sdks/ruby/test/test_client.rb
require_relative 'test_helper'

class TestClient < Minitest::Test
  def setup
    @server = MockServer.new
    @client = Beachcomber::Client.new(socket_path: @server.path)
  end

  def teardown
    @server.stop
  end

  # --- get ---

  def test_get_hit
    @server.enqueue('{"ok":true,"data":"main","age_ms":42,"stale":false}')
    result = @client.get('git.branch')
    assert result.hit?
    assert_equal 'main', result.data
    assert_equal 42,     result.age_ms
    refute result.stale?
  end

  def test_get_miss
    @server.enqueue('{"ok":true}')
    result = @client.get('git.branch')
    assert result.miss?
    assert_nil result.data
  end

  def test_get_null_data_treated_as_miss
    @server.enqueue('{"ok":true,"data":null}')
    result = @client.get('git.branch')
    assert result.miss?
  end

  def test_get_with_path
    @server.enqueue('{"ok":true,"data":"main","age_ms":5,"stale":false}')
    @client.get('git.branch', path: '/repo')
    req = @server.requests.last
    assert_equal 'get',    req['op']
    assert_equal 'git.branch', req['key']
    assert_equal '/repo',  req['path']
  end

  def test_get_without_path_omits_field
    @server.enqueue('{"ok":true}')
    @client.get('git.branch')
    req = @server.requests.last
    refute req.key?('path')
  end

  def test_get_hash_data
    @server.enqueue('{"ok":true,"data":{"branch":"main","dirty":false},"age_ms":1,"stale":false}')
    result = @client.get('git')
    assert result.hit?
    assert_equal 'main',  result['branch']
    assert_equal false,   result['dirty']
  end

  def test_get_stale_flag
    @server.enqueue('{"ok":true,"data":"old","age_ms":9000,"stale":true}')
    result = @client.get('git.branch')
    assert result.stale?
  end

  # --- server error ---

  def test_get_server_error_raises
    @server.enqueue('{"ok":false,"error":"unknown provider: foo"}')
    err = assert_raises(Beachcomber::ServerError) { @client.get('foo.bar') }
    assert_match(/unknown provider: foo/, err.message)
  end

  # --- refresh ---

  def test_refresh_sends_correct_op
    @server.enqueue('{"ok":true}')
    @client.refresh('git')
    req = @server.requests.last
    assert_equal 'refresh', req['op']
    assert_equal 'git',     req['key']
  end

  def test_refresh_with_path
    @server.enqueue('{"ok":true}')
    @client.refresh('git', path: '/repo')
    req = @server.requests.last
    assert_equal '/repo', req['path']
  end

  def test_refresh_returns_nil
    @server.enqueue('{"ok":true}')
    result = @client.refresh('git')
    assert_nil result
  end

  # --- status ---

  def test_status_sends_op
    @server.enqueue('{"ok":true,"data":[]}')
    rows = @client.status
    assert_equal [], rows
    assert_equal 'status', @server.requests.last['op']
  end

  # --- connection error ---

  # An explicit-but-unreachable socket_path surfaces as
  # Beachcomber::ConnectionFailedError (the shared client's
  # CombError::ConnectionFailed) rather than DaemonNotRunning, which is
  # reserved for the no-override, autostart-disabled default-discovery path.
  # Both share the CombError ancestor and carry a `kind`.
  def test_daemon_not_running_when_no_socket
    client = Beachcomber::Client.new(socket_path: '/tmp/beachcomber-no-such-socket-xyz/sock')
    err = assert_raises(Beachcomber::CombError) { client.get('git.branch') }
    assert_includes %w[connection_failed daemon_not_running], err.kind
  end

  # --- protocol errors ---

  def test_malformed_json_raises_protocol_error
    @server.enqueue('not json at all')
    assert_raises(Beachcomber::ProtocolError) { @client.get('git.branch') }
  end

  # --- session ---

  def test_session_block_yields_session
    @server.enqueue('{"ok":true,"data":"main","age_ms":1,"stale":false}')
    @client.session do |s|
      assert_instance_of Beachcomber::Session, s
      r = s.get('git.branch')
      assert_equal 'main', r.data
    end
  end

  def test_session_set_context
    @server.enqueue('{"ok":true}') # context ack
    @server.enqueue('{"ok":true,"data":"main","age_ms":1,"stale":false}')

    @client.session do |s|
      s.set_context('/repo')
      r = s.get('git.branch')
      assert r.hit?
    end

    ctx_req = @server.requests.first
    assert_equal 'context', ctx_req['op']
    assert_equal '/repo',   ctx_req['path']
  end

  def test_session_multiple_requests_on_one_connection
    3.times { @server.enqueue('{"ok":true,"data":"main","age_ms":1,"stale":false}') }

    @client.session do |s|
      3.times { s.get('git.branch') }
    end

    assert_equal 3, @server.requests.size
  end

  def test_session_closes_on_exception
    @server.enqueue('{"ok":false,"error":"oops"}')
    assert_raises(Beachcomber::ServerError) do
      @client.session { |s| s.get('git.branch') }
    end
    # The server should still accept a new connection (connection was closed).
    @server.enqueue('{"ok":true,"data":"x","age_ms":0,"stale":false}')
    result = @client.get('git.branch')
    assert result.hit?
  end

  def test_session_returns_block_value
    @server.enqueue('{"ok":true,"data":"main","age_ms":1,"stale":false}')
    value = @client.session { |s| s.get('git.branch').data }
    assert_equal 'main', value
  end

  # --- each op sends correct JSON ---

  def test_get_op_field
    @server.enqueue('{"ok":true}')
    @client.get('hostname')
    assert_equal 'get', @server.requests.last['op']
  end

  def test_status_op_field
    @server.enqueue('{"ok":true,"data":[]}')
    @client.status
    assert_equal 'status', @server.requests.last['op']
  end
end

File: sdks/lua/test/test_client.lua
--- Tests for beachcomber.client: Result accessors and Client against a
--- mock backend (the interface documented at the top of
--- beachcomber/client.lua — canonical response tables, not raw sockets).

return function(suite, test, skip, assert_eq, assert_true, assert_nil, assert_not_nil)
  local client_mod = require("beachcomber.client")
  local Client = client_mod.Client
  local Result = client_mod.Result

  -- ── Result accessors ──────────────────────────────────────────────────────

  suite("Result accessors")

  test("is_hit returns false on miss", function()
    local r = Result.new({ data = nil })
    assert_true(not r:is_hit(), "nil data should be a miss")
    assert_nil(r.data)
    assert_eq(r.age_ms, 0)
    assert_eq(r.stale, false)
  end)

  test("is_hit returns true on hit", function()
    local r = Result.new({ data = "main", age_ms = 500, stale = false })
    assert_true(r:is_hit())
    assert_eq(r.data, "main")
    assert_eq(r.age_ms, 500)
  end)

  test("is_hit true for numeric 0 data", function()
    local r = Result.new({ data = 0, age_ms = 10, stale = false })
    assert_true(r:is_hit())
  end)

  test("is_hit true for boolean false data", function()
    local r = Result.new({ data = false, age_ms = 10, stale = false })
    assert_true(r:is_hit())
  end)

  test("get_str on object data", function()
    local r = Result.new({ data = { branch = "main", dirty = false }, age_ms = 100 })
    local v, err = r:get_str("branch")
    assert_nil(err)
    assert_eq(v, "main")
  end)

  test("get_str missing field", function()
    local r = Result.new({ data = { branch = "main" } })
    local v, err = r:get_str("dirty")
    assert_nil(v)
    assert_not_nil(err)
  end)

  test("get_str on non-object data errors", function()
    local r = Result.new({ data = "main" })
    local v, err = r:get_str("anything")
    assert_nil(v)
    assert_not_nil(err)
  end)

  test("stale defaults to false", function()
    local r = Result.new({ data = "main", age_ms = 1000 })
    assert_eq(r.stale, false)
  end)

  -- ── Mock backend helper ────────────────────────────────────────────────────

  local function make_backend(overrides)
    local backend = { _calls = {} }
    backend.name = function(_) return "mock" end
    for k, v in pairs(overrides or {}) do backend[k] = v end
    return backend
  end

  -- ── Client:get / get_with_flags ───────────────────────────────────────────

  suite("Client:get")

  test("get returns Result on hit", function()
    local backend = make_backend({
      get = function(_, key, path, flags)
        return { ok = true, data = "main", age_ms = 500, stale = false }
      end,
    })
    local c = Client.new(backend)
    local r, err = c:get("git.branch", "/repo")
    assert_nil(err)
    assert_true(r:is_hit())
    assert_eq(r.data, "main")
  end)

  test("get forwards key and path to the backend", function()
    local seen
    local backend = make_backend({
      get = function(_, key, path, flags)
        seen = { key = key, path = path }
        return { ok = true }
      end,
    })
    Client.new(backend):get("git.branch", "/myrepo")
    assert_eq(seen.key, "git.branch")
    assert_eq(seen.path, "/myrepo")
  end)

  test("get returns miss Result when data is nil", function()
    local backend = make_backend({ get = function() return { ok = true } end })
    local r = Client.new(backend):get("git.branch")
    assert_true(not r:is_hit())
  end)

  test("get returns nil+Error when backend returns ok=false", function()
    local backend = make_backend({
      get = function() return { ok = false, error = { kind = "server_error", message = "unknown provider: bad" } } end,
    })
    local r, err = Client.new(backend):get("bad.field")
    assert_nil(r)
    assert_not_nil(err)
    assert_eq(err.kind, "server_error")
    assert_true(tostring(err):find("unknown provider", 1, true) ~= nil)
  end)

  suite("Client:get_with_flags")

  test("get_with_flags propagates force/wait as a flags table", function()
    local seen_flags
    local backend = make_backend({
      get = function(_, key, path, flags) seen_flags = flags; return { ok = true } end,
    })
    Client.new(backend):get_with_flags("git.branch", nil, true, nil)
    assert_eq(seen_flags.force, true)
    assert_true(not seen_flags.wait)
  end)

  -- ── Client:refresh / set_context ──────────────────────────────────────────

  suite("Client:refresh")

  test("refresh returns true on success and forwards key/path", function()
    local seen
    local backend = make_backend({
      refresh = function(_, key, path) seen = { key = key, path = path }; return { ok = true } end,
    })
    local ok, err = Client.new(backend):refresh("git", "/repo")
    assert_nil(err)
    assert_true(ok)
    assert_eq(seen.key, "git")
    assert_eq(seen.path, "/repo")
  end)

  suite("Client:set_context")

  test("set_context supplies the default path to a later call with no explicit path", function()
    local seen_path
    local backend = make_backend({
      get = function(_, key, path) seen_path = path; return { ok = true } end,
    })
    local c = Client.new(backend)
    c:set_context("/some/dir")
    c:get("git.branch")
    assert_eq(seen_path, "/some/dir")
  end)

  test("an explicit path argument overrides the context default", function()
    local seen_path
    local backend = make_backend({
      get = function(_, key, path) seen_path = path; return { ok = true } end,
    })
    local c = Client.new(backend)
    c:set_context("/context/dir")
    c:get("git.branch", "/explicit/dir")
    assert_eq(seen_path, "/explicit/dir")
  end)

  -- ── Client:status ─────────────────────────────────────────────────────────

  suite("Client:status")

  test("status returns data array", function()
    local backend = make_backend({
      status = function() return { ok = true, data = { { key = "git.branch" }, { key = "git.dirty" } } } end,
    })
    local result, err = Client.new(backend):status()
    assert_nil(err)
    assert_eq(#result, 2)
  end)

  test("status returns empty array when data is nil", function()
    local backend = make_backend({ status = function() return { ok = true } end })
    local result = Client.new(backend):status()
    assert_eq(#result, 0)
  end)

  test("status returns nil+Error on backend error", function()
    local backend = make_backend({
      status = function() return { ok = false, error = { kind = "io_error", message = "internal error" } } end,
    })
    local result, err = Client.new(backend):status()
    assert_nil(result)
    assert_not_nil(err)
  end)

  -- ── Client:transport ─────────────────────────────────────────────────────

  suite("Client:transport")

  test("transport reports the backend's name", function()
    local backend = make_backend({ name = function() return "ffi" end })
    assert_eq(Client.new(backend):transport(), "ffi")
  end)
end
